### PR TITLE
Add docker support

### PR DESCRIPTION
**Pull request by dgreene-r7**
This PR adds basic support for the Docker builder and the Docker post-processors.

### CODE REVIEW

File: docs/configuration/profile.md [hunk 1]
```diff
@@ -29,9 +29,22 @@ Packer configurations for this profile
...
+Options for the `docker` builder:
+* `image` The base image for the Docker container that will be started
+
+The Docker builder requires one, and only one, of the following options:
+* `commit` The container will be committed to an image rather than exported
+* `discard` Throw away the container when the build is complete
+* `export_path` The path where the final container will be exported as a tar file
+
+There are additional options specified in `lib/builderator/config/file.rb` and
+in the Packer documentation.
```
Comment (erran-r7) [inline docs/configuration/profile.md:47]: 🔗 ?

File: docs/configuration/profile.md [hunk 2]
```diff
@@ -87,6 +100,37 @@ Add a packer build
...
+* `repository` The repository of the imported image
+* `tag` The tag for the imported image
+* `force` If true, the `docker-tag` post-processor forcibly
+tags the image even if there is a tag name collision. Defaults to `false`.
+
+[Docker-save][] post-processor
+* `path` The path to save the image
+
+[Docker-push][] post-processor
+
```
Comment (erran-r7) [inline docs/configuration/profile.md:125]: The above types do not have a newline between the bulleted list.

File: lib/builderator/config/defaults.rb [hunk 3]
```diff
@@ -80,6 +80,16 @@ module Config
         end
       end
 
+      profile :docker do |profile|
+        profile.log_level :info
+
+        profile.packer do |packer|
+          packer.build :docker do |build|
+            build.type 'docker'
```
Comment (erran-r7) [inline lib/builderator/config/defaults.rb:88]: Are there any defaults we want to add here like we have for the `amazon-ebs`/`virtualbox` types?

File: lib/builderator/interface/packer.rb [hunk 4]
```diff
@@ -21,11 +21,21 @@ def initialize(*_)
...
             build_hash = build.to_hash.tap do |b|
-              b[:tags] = Config.profile.current.tags
+              b[:tags] = Config.profile.current.tags unless Config.profile.current.tags.empty?
+            end
+
+            if build_hash[:type] == 'docker'
+              raise 'The Docker builder requires a base image' unless build_hash.key?(:image)
+
+              # The Docker builder requires one and only one of 'commit', 'discard', or 'export_path' set
+              if build_hash.keys.find_all { |k| [:commit, :discard, :export_path].include?(k) }.length != 1
```
Comment (erran-r7) [inline lib/builderator/interface/packer.rb]: **stylistic**: Do we use `#find_all` or `#select` in this repository as per the [ruby style guide prefer select rule](https://github.com/bbatsov/ruby-style-guide#map-find-select-reduce-size)?

File: lib/builderator/interface/packer.rb [hunk 5]
```diff
@@ -49,12 +59,33 @@ def initialize(*_)
...
+              post_processors << name
+              next
+            end
+
+            # The post-processor's type should be the same as the name
+            post_processor_hash[:type] = name
+
+            post_processors << post_processor_hash
+          end
+          json['post-processors'].push(post_processors)
```
Comment (erran-r7) [inline lib/builderator/interface/packer.rb:83]: Should this be a flat array of post processors? Right now it generates `[[...]]` right?

```ruby
['foo', 'bar'].push(['baz', 'qux'])
#=> ["foo", "bar", ["baz", "qux"]]
```

File: lib/builderator/interface/packer.rb (hunk 4 above)
Comment (dgreene-r7) [reply lib/builderator/interface/packer.rb]: Good catch. I was looking at the difference between `#find_all` and `#select` as it applies to `Hash` even though `Hash.keys` returns an `Array`.

Fixed in efbdb6f.

File: lib/builderator/interface/packer.rb (hunk 5 above)
Comment (dgreene-r7) [reply lib/builderator/interface/packer.rb:83]: There's 3 different types of post-processors: a simple definition, a detailed definition, and a sequence definition.

There's no penalty for always using a sequence definition (which is what we're doing) versus including logic to determine if we should use simple, detailed, or sequences.

See https://www.packer.io/docs/templates/post-processors.html for more info.

File: lib/builderator/config/defaults.rb (hunk 3 above)
Comment (dgreene-r7) [reply lib/builderator/config/defaults.rb:88]: Not that I can see. All other attributes either have Packer defaults or should be specified in the Buildfile.

**Review by fmitchell-r7 — APPROVED** [review]

**Review by erran-r7 — APPROVED** [review]

**Pull request merged**